File: frontend/src/shared/lib/messageLinks.js
const MESSAGE_LINK_REGEX = /https?:\/\/[^\s<>()]+|(?:www\.)?(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,}(?:\/[^\s<>()]*)?/gi;
const TRAILING_PUNCTUATION = new Set(['.', ',', '!', '?', ':', ';', ')', ']', '}', '"', '\'', '»']);

function trimTrailingPunctuation(value) {
  let end = value.length;
  while (end > 0 && TRAILING_PUNCTUATION.has(value[end - 1])) {
    end -= 1;
  }
  return value.slice(0, end);
}

function normalizeLinkHref(value) {
  const trimmed = String(value || '').trim();
  if (!trimmed) return '';
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  return `https://${trimmed}`;
}

export function parseMessageTextParts(text) {
  const source = String(text || '');
  if (!source) return [];

  const parts = [];
  let lastIndex = 0;

  for (const match of source.matchAll(MESSAGE_LINK_REGEX)) {
    const rawMatch = String(match[0] || '');
    const matchIndex = Number(match.index);
    if (!rawMatch || !Number.isInteger(matchIndex) || matchIndex < lastIndex) continue;

    const previousChar = source[matchIndex - 1] || '';
    if (previousChar === '@') continue;

    const visibleText = trimTrailingPunctuation(rawMatch);
    if (!visibleText) continue;

    const normalizedHref = normalizeLinkHref(visibleText);
    const visibleEnd = matchIndex + visibleText.length;

    if (matchIndex > lastIndex) {
      parts.push({
        type: 'text',
        text: source.slice(lastIndex, matchIndex)
      });
    }

    parts.push({
      type: 'link',
      text: visibleText,
      href: normalizedHref
    });

    lastIndex = visibleEnd;
  }

  if (lastIndex < source.length) {
    parts.push({
      type: 'text',
      text: source.slice(lastIndex)
    });
  }

  if (parts.length === 0) {
    return [{ type: 'text', text: source }];
  }

  return parts;
}

File: android-app/app/src/main/java/ru/govchat/app/ui/screens/main/MessageLinks.kt
package ru.govchat.app.ui.screens.main

private val MESSAGE_LINK_REGEX = Regex(
    pattern = """(?i)https?://[^\s<>()]+|(?:www\.)?(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,}(?:/[^\s<>()]*)?"""
)

private val TRAILING_PUNCTUATION = setOf('.', ',', '!', '?', ':', ';', ')', ']', '}', '"', '\'', '»')

data class MessageTextPart(
    val text: String,
    val url: String? = null
)

fun parseMessageTextParts(text: String?): List<MessageTextPart> {
    val source = text.orEmpty()
    if (source.isEmpty()) return emptyList()

    val parts = mutableListOf<MessageTextPart>()
    var lastIndex = 0

    MESSAGE_LINK_REGEX.findAll(source).forEach { match ->
        val rawMatch = match.value
        val startIndex = match.range.first
        if (startIndex < lastIndex) return@forEach

        val previousChar = source.getOrNull(startIndex - 1)
        if (previousChar == '@') return@forEach

        val visibleText = trimTrailingPunctuation(rawMatch)
        if (visibleText.isEmpty()) return@forEach

        val visibleEnd = startIndex + visibleText.length
        if (startIndex > lastIndex) {
            parts += MessageTextPart(text = source.substring(lastIndex, startIndex))
        }

        parts += MessageTextPart(
            text = visibleText,
            url = normalizeLinkUrl(visibleText)
        )
        lastIndex = visibleEnd
    }

    if (lastIndex < source.length) {
        parts += MessageTextPart(text = source.substring(lastIndex))
    }

    return if (parts.isEmpty()) {
        listOf(MessageTextPart(text = source))
    } else {
        parts
    }
}

private fun trimTrailingPunctuation(value: String): String {
    var endIndex = value.length
    while (endIndex > 0 && value[endIndex - 1] in TRAILING_PUNCTUATION) {
        endIndex -= 1
    }
    return value.substring(0, endIndex)
}

private fun normalizeLinkUrl(value: String): String {
    val trimmed = value.trim()
    if (trimmed.isEmpty()) return ""
    return if (trimmed.startsWith("http://", ignoreCase = true) || trimmed.startsWith("https://", ignoreCase = true)) {
        trimmed
    } else {
        "https://$trimmed"
    }
}
